os/drivers/kdbg/kdbg_drv.c

```diff
@@ -0,0 +1,155 @@
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <debug.h>
+#if CONFIG_TASK_NAME_SIZE > 0
+#include <string.h>
+#endif
+#include <tinyara/sched.h>
+#include <tinyara/kdbg_drv.h>
+#include <tinyara/mm/mm.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/fs/ioctl.h>
+
+#ifdef CONFIG_MM_KERNEL_HEAP
+extern struct mm_heap_s g_kmmheap;
+#endif
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int kdbg_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
+static ssize_t kdbg_read(FAR struct file *filep, FAR char *buffer, size_t len);
+static ssize_t kdbg_write(FAR struct file *filep, FAR const char *buffer, size_t len);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct kdbg_dev_s {
+	mqd_t mqdes;	   /* A mqueue descriptor */
+};
+
+static const struct file_operations kdbg_fops = {
+	0,                          /* open */
+	0,                          /* close */
+	kdbg_read,               /* read */
+	kdbg_write,              /* write */
+	0,                          /* seek */
+	kdbg_ioctl               /* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	, 0                         /* poll */
+#endif
+};
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static ssize_t kdbg_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+	return 0;
+}
+
+static ssize_t kdbg_write(FAR struct file *filep, FAR const char *buffer, size_t len)
+{
+	return 0;
+}
+
+/************************************************************************************
+ * Name: kdbg_ioctl
+ *
+ * Description: The ioctl method for kdbg.
+ *
+ ************************************************************************************/
+static int kdbg_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = ERROR;
+	struct tcb_s *tcb;
+#ifdef CONFIG_MM_KERNEL_HEAP
+	heapinfo_option_t *option;
+#endif
+#if CONFIG_TASK_NAME_SIZE > 0
+	heapinfo_taskname_t *task_info;
+#endif
+
+	/* Handle built-in ioctl commands */
+
+	switch (cmd) {
+#ifdef CONFIG_MM_KERNEL_HEAP
+	case KDBGIOC_HEAPINFO_PARSE:
+		option = (heapinfo_option_t *)arg;
+		heapinfo_parse(&g_kmmheap, option->mode, option->pid);
+		ret = OK;
+		break;
+#if CONFIG_MM_REGIONS > 1
+	case KDBGIOC_HEAPINFO_NREGION:
+		ret = g_kmmheap.mm_nregions;
+		break;
+#endif
+#endif
+#if defined(CONFIG_SCHED_HAVE_PARENT) && !defined(HAVE_GROUP_MEMBERS)
+	case KDBGIOC_HEAPINFO_PPID:
```

Comment:
PPID, STACKSIZE and TASKNAME do not have any relationship with heapinfo.
You can use them from heapinfo but it does not mean they are information of heapinfo.
Let's use a way of procfs which @jeongarmy made.